Exception occurs in v0.9.1 using embulk-filter-add_time

In v.0.9.1 using embulk-filter-add_time, requiring TimeZone Parameter but it could not find a configuration point. And this does not occur in v.0.9.0 so it may have some breaking changes. Is this bug?

- https://github.com/treasure-data/embulk-filter-add_time


### config.yml
```
in:
  type: file
  path_prefix: csv/sample_
  decoders:
  - {type: gzip}
  parser:
    charset: UTF-8
    newline: LF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '"'
    null_string: 'NULL'
    trim_if_not_quoted: false
    skip_header_lines: 1
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: id, type: long}
    - {name: account, type: long}
    - {name: time, type: timestamp, format: '%Y-%m-%d %H:%M:%S'}
    - {name: purchase, type: timestamp, format: '%Y%m%d'}
    - {name: comment, type: string}
filters:
- type: add_time
  to_column:
    name: time
    type: timestamp
  from_value:
    value: "2016-01-01 00:00:00 UTC"
  TimeZone: UTC
out: {type: stdout}
```

### StackTrace 
```
bash-3.2$ embulk preview config.yml
2018-02-10 22:38:11.129 +0900: Embulk v0.9.1

********************************** INFORMATION **********************************
  Join us! Embulk-announce mailing list is up for IMPORTANT announcement such as
    compatibility-breaking changes and key feature updates.
  https://groups.google.com/forum/#!forum/embulk-announce
*********************************************************************************

2018-02-10 22:38:12.917 +0900 [INFO] (main): Gem's home and path are set by default: "/Users/smdmts/.embulk/lib/gems"
2018-02-10 22:38:13.397 +0900 [INFO] (main): Started Embulk v0.9.1
2018-02-10 22:38:13.460 +0900 [INFO] (0001:preview): Loaded plugin embulk-filter-add_time (0.2.0)
2018-02-10 22:38:13.488 +0900 [INFO] (0001:preview): Listing local files at directory 'csv' filtering filename by prefix 'sample_'
2018-02-10 22:38:13.489 +0900 [INFO] (0001:preview): "follow_symlinks" is set false. Note that symbolic links to directories are skipped.
2018-02-10 22:38:13.493 +0900 [INFO] (0001:preview): Loading files [csv/sample_01.csv.gz]
2018-02-10 22:38:13.501 +0900 [INFO] (0001:preview): Try to read 32,768 bytes from input source
2018-02-10 22:38:13.576 +0900 [WARN] (0001:preview): to_column 'time' is set but 'time' column also exists. The existent 'time' column is renamed to 'time_'.
java.lang.IllegalArgumentException: Field 'TimeZone' is required but not set
 at [Source: N/A; line: -1, column: -1]
    at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3459)
    at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:3400)
    at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:3388)
    at org.embulk.config.TaskSerDe$TaskDeserializer.deserialize(TaskSerDe.java:152)
    at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3708)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2005)
    at org.embulk.config.ModelManager.readObject(ModelManager.java:45)
    at org.embulk.config.DataSourceImpl.loadTask(DataSourceImpl.java:192)
    at org.embulk.filter.add_time.AddTimeFilterPlugin.open(AddTimeFilterPlugin.java:168)
    at org.embulk.spi.util.Filters.open(Filters.java:50)
    at org.embulk.exec.PreviewExecutor$2$1.run(PreviewExecutor.java:127)
    at org.embulk.spi.util.Filters$RecursiveControl.transaction(Filters.java:84)
    at org.embulk.spi.util.Filters$RecursiveControl$1.run(Filters.java:80)
    at org.embulk.filter.add_time.AddTimeFilterPlugin.transaction(AddTimeFilterPlugin.java:161)
    at org.embulk.spi.util.Filters$RecursiveControl.transaction(Filters.java:76)
    at org.embulk.spi.util.Filters.transaction(Filters.java:42)
    at org.embulk.exec.PreviewExecutor$2.run(PreviewExecutor.java:118)
    at org.embulk.spi.FileInputRunner$RunnerControl$1$1.run(FileInputRunner.java:112)
    at org.embulk.standards.CsvParserPlugin.transaction(CsvParserPlugin.java:228)
    at org.embulk.spi.FileInputRunner$RunnerControl$1.run(FileInputRunner.java:107)
    at org.embulk.spi.util.Decoders$RecursiveControl.transaction(Decoders.java:68)
    at org.embulk.spi.util.Decoders$RecursiveControl$1.run(Decoders.java:64)
    at org.embulk.standards.GzipFileDecoderPlugin.transaction(GzipFileDecoderPlugin.java:25)
    at org.embulk.spi.util.Decoders$RecursiveControl.transaction(Decoders.java:60)
    at org.embulk.spi.util.Decoders.transaction(Decoders.java:29)
    at org.embulk.spi.FileInputRunner$RunnerControl.run(FileInputRunner.java:105)
    at org.embulk.exec.BufferFileInputPlugin.transaction(BufferFileInputPlugin.java:21)
    at org.embulk.spi.FileInputRunner.transaction(FileInputRunner.java:62)
    at org.embulk.exec.PreviewExecutor.doPreview(PreviewExecutor.java:116)
    at org.embulk.exec.PreviewExecutor.doPreview(PreviewExecutor.java:102)
    at org.embulk.exec.PreviewExecutor.access$000(PreviewExecutor.java:30)
    at org.embulk.exec.PreviewExecutor$1.run(PreviewExecutor.java:74)
    at org.embulk.exec.PreviewExecutor$1.run(PreviewExecutor.java:71)
    at org.embulk.spi.Exec.doWith(Exec.java:22)
    at org.embulk.exec.PreviewExecutor.preview(PreviewExecutor.java:71)
    at org.embulk.EmbulkEmbed.preview(EmbulkEmbed.java:152)
    at org.embulk.EmbulkRunner.previewInternal(EmbulkRunner.java:214)
    at org.embulk.EmbulkRunner.preview(EmbulkRunner.java:106)
    at org.embulk.cli.EmbulkRun.runSubcommand(EmbulkRun.java:433)
    at org.embulk.cli.EmbulkRun.run(EmbulkRun.java:91)
    at org.embulk.cli.Main.main(Main.java:26)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Field 'TimeZone' is required but not set
 at [Source: N/A; line: -1, column: -1]
    at org.embulk.config.TaskSerDe$TaskDeserializer.deserialize(TaskSerDe.java:180)
    at com.fasterxml.jackson.datatype.guava.deser.GuavaOptionalDeserializer.deserialize(GuavaOptionalDeserializer.java:118)
    at com.fasterxml.jackson.datatype.guava.deser.GuavaOptionalDeserializer.deserialize(GuavaOptionalDeserializer.java:16)
    at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3454)
    ... 40 more

Error: Field 'TimeZone' is required but not set
 at [Source: N/A; line: -1, column: -1]
bash-3.2$ embulk preview config.yml
2018-02-10 22:39:04.763 +0900: Embulk v0.9.1

```

```
2018-02-10 22:47:36.437 +0900: Embulk v0.9.0

********************************** INFORMATION **********************************
  Join us! Embulk-announce mailing list is up for IMPORTANT announcement such as
    compatibility-breaking changes and key feature updates.
  https://groups.google.com/forum/#!forum/embulk-announce
*********************************************************************************

2018-02-10 22:47:38.277 +0900 [INFO] (main): Gem's home and path are set by default: "/Users/smdmts/.embulk/lib/gems"
2018-02-10 22:47:38.863 +0900 [INFO] (main): Started Embulk v0.9.0
2018-02-10 22:47:38.926 +0900 [INFO] (0001:preview): Loaded plugin embulk-filter-add_time (0.2.0)
2018-02-10 22:47:38.950 +0900 [INFO] (0001:preview): Listing local files at directory 'csv' filtering filename by prefix 'sample_'
2018-02-10 22:47:38.951 +0900 [INFO] (0001:preview): "follow_symlinks" is set false. Note that symbolic links to directories are skipped.
2018-02-10 22:47:38.956 +0900 [INFO] (0001:preview): Loading files [csv/sample_01.csv.gz]
2018-02-10 22:47:38.963 +0900 [INFO] (0001:preview): Try to read 32,768 bytes from input source
2018-02-10 22:47:39.044 +0900 [WARN] (0001:preview): to_column 'time' is set but 'time' column also exists. The existent 'time' column is renamed to 'time_'.
2018-02-10 22:47:39.058 +0900 [WARN] (0001:preview): to_column 'time' is set but 'time' column also exists. The existent 'time' column is renamed to 'time_'.
+---------+--------------+-------------------------+-------------------------+----------------------------+-------------------------+
| id:long | account:long |         time_:timestamp |      purchase:timestamp |             comment:string |          time:timestamp |
+---------+--------------+-------------------------+-------------------------+----------------------------+-------------------------+
|       1 |       32,864 | 2015-01-27 19:23:49 UTC | 2015-01-27 00:00:00 UTC |                     embulk | 2016-01-01 00:00:00 UTC |
|       2 |       14,824 | 2015-01-27 19:01:23 UTC | 2015-01-27 00:00:00 UTC |               embulk jruby | 2016-01-01 00:00:00 UTC |
|       3 |       27,559 | 2015-01-28 02:20:02 UTC | 2015-01-28 00:00:00 UTC | Embulk "csv" parser plugin | 2016-01-01 00:00:00 UTC |
|       4 |       11,270 | 2015-01-29 11:54:36 UTC | 2015-01-29 00:00:00 UTC |                            | 2016-01-01 00:00:00 UTC |
+---------+--------------+-------------------------+-------------------------+----------------------------+-------------------------+
```

--- a/embulk-core/src/main/java/org/embulk/config/TaskInvocationHandler.java
+++ b/embulk-core/src/main/java/org/embulk/config/TaskInvocationHandler.java
@@ -2,6 +2,8 @@ package org.embulk.config;
 
 import com.google.common.collect.ImmutableMultimap;
 import com.google.common.collect.Multimap;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.Method;
 import java.lang.reflect.Proxy;
@@ -32,7 +34,9 @@ class TaskInvocationHandler implements InvocationHandler {
         for (Method method : iface.getMethods()) {
             String methodName = method.getName();
             String fieldName = getterFieldNameOrNull(methodName);
-            if (fieldName != null && hasExpectedArgumentLength(method, 0)) {
+            if (fieldName != null && hasExpectedArgumentLength(method, 0)
+                    && (!method.isDefault() || method.getAnnotation(Config.class) != null)) {
+                // If the method has default implementation, and @Config is not annotated there, the method is kept.
                 builder.put(fieldName, method);
             }
         }
@@ -122,6 +126,52 @@ class TaskInvocationHandler implements InvocationHandler {
                 String fieldName;
                 fieldName = getterFieldNameOrNull(methodName);
                 if (fieldName != null) {
+                    if (method.isDefault() && !this.objects.containsKey(fieldName)) {
+                        // If and only if the method has default implementation, and @Config is not annotated there,
+                        // it is tried to call the default implementation directly without proxying.
+                        //
+                        // methodWithDefaultImpl.invoke(proxy) without this hack would cause infinite recursive calls.
+                        //
+                        // See hints:
+                        // https://rmannibucau.wordpress.com/2014/03/27/java-8-default-interface-methods-and-jdk-dynamic-proxies/
+                        // https://stackoverflow.com/questions/22614746/how-do-i-invoke-java-8-default-methods-reflectively
+                        //
+                        // This hack is required to support `org.joda.time.DateTimeZone` in some Tasks, for example
+                        // TimestampParser.Task and TimestampParser.TimestampColumnOption.
+                        //
+                        // TODO: Remove the hack once a cleaner way is found, or Joda-Time is finally removed.
+                        // https://github.com/embulk/embulk/issues/890
+                        if (CONSTRUCTOR_MethodHandles_Lookup != null) {
+                            synchronized (CONSTRUCTOR_MethodHandles_Lookup) {
+                                boolean hasSetAccessible = false;
+                                try {
+                                    CONSTRUCTOR_MethodHandles_Lookup.setAccessible(true);
+                                    hasSetAccessible = true;
+                                } catch (SecurityException ex) {
+                                    // Skip handling default implementation in case of errors.
+                                }
+
+                                if (hasSetAccessible) {
+                                    try {
+                                        return CONSTRUCTOR_MethodHandles_Lookup
+                                                .newInstance(
+                                                        method.getDeclaringClass(),
+                                                        MethodHandles.Lookup.PUBLIC
+                                                                | MethodHandles.Lookup.PRIVATE
+                                                                | MethodHandles.Lookup.PROTECTED
+                                                                | MethodHandles.Lookup.PACKAGE)
+                                                .unreflectSpecial(method, method.getDeclaringClass())
+                                                .bindTo(proxy)
+                                                .invokeWithArguments();
+                                    } catch (Throwable ex) {
+                                        // Skip handling default implementation in case of errors.
+                                    } finally {
+                                        CONSTRUCTOR_MethodHandles_Lookup.setAccessible(false);
+                                    }
+                                }
+                            }
+                        }
+                    }
                     checkArgumentLength(method, 0, methodName);
                     return invokeGetter(method, fieldName);
                 }
@@ -161,4 +211,18 @@ class TaskInvocationHandler implements InvocationHandler {
                     String.format("Method '%s' expected %d argument but got %d arguments", methodName, expected, method.getParameterTypes().length));
         }
     }
+
+    static {
+        Constructor<MethodHandles.Lookup> constructorMethodHandlesLookupTemporary = null;
+        try {
+            constructorMethodHandlesLookupTemporary =
+                    MethodHandles.Lookup.class.getDeclaredConstructor(Class.class, int.class);
+        } catch (NoSuchMethodException | SecurityException ex) {
+            constructorMethodHandlesLookupTemporary = null;
+        } finally {
+            CONSTRUCTOR_MethodHandles_Lookup = constructorMethodHandlesLookupTemporary;
+        }
+    }
+
+    private static final Constructor<MethodHandles.Lookup> CONSTRUCTOR_MethodHandles_Lookup;
 }

--- a/embulk-core/src/main/java/org/embulk/spi/time/TimestampFormatter.java
+++ b/embulk-core/src/main/java/org/embulk/spi/time/TimestampFormatter.java
@@ -104,9 +104,13 @@ public class TimestampFormatter {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("default_timezone")
-        @ConfigDefault("\"UTC\"")
-        public org.joda.time.DateTimeZone getDefaultTimeZone();
+        public default org.joda.time.DateTimeZone getDefaultTimeZone() {
+            if (getDefaultTimeZoneId() != null) {
+                return TimeZoneIds.parseJodaDateTimeZone(getDefaultTimeZoneId());
+            } else {
+                return null;
+            }
+        }
 
         @Config("default_timestamp_format")
         @ConfigDefault("\"%Y-%m-%d %H:%M:%S.%6N %z\"")
@@ -121,9 +125,13 @@ public class TimestampFormatter {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("timezone")
-        @ConfigDefault("null")
-        public Optional<org.joda.time.DateTimeZone> getTimeZone();
+        public default Optional<org.joda.time.DateTimeZone> getTimeZone() {
+            if (getTimeZoneId().isPresent()) {
+                return Optional.of(TimeZoneIds.parseJodaDateTimeZone(getTimeZoneId().get()));
+            } else {
+                return Optional.absent();
+            }
+        }
 
         @Config("format")
         @ConfigDefault("null")

--- a/embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java
+++ b/embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java
@@ -164,9 +164,13 @@ public class TimestampParser {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("default_timezone")
-        @ConfigDefault("\"UTC\"")
-        public org.joda.time.DateTimeZone getDefaultTimeZone();
+        public default org.joda.time.DateTimeZone getDefaultTimeZone() {
+            if (getDefaultTimeZoneId() != null) {
+                return TimeZoneIds.parseJodaDateTimeZone(getDefaultTimeZoneId());
+            } else {
+                return null;
+            }
+        }
 
         @Config("default_timestamp_format")
         @ConfigDefault("\"%Y-%m-%d %H:%M:%S.%N %z\"")
@@ -185,7 +189,13 @@ public class TimestampParser {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        public Optional<org.joda.time.DateTimeZone> getTimeZone();
+        public default Optional<org.joda.time.DateTimeZone> getTimeZone() {
+            if (getTimeZoneId().isPresent()) {
+                return Optional.of(TimeZoneIds.parseJodaDateTimeZone(getTimeZoneId().get()));
+            } else {
+                return Optional.absent();
+            }
+        }
 
         @Config("format")
         @ConfigDefault("null")

--- a/embulk-core/src/main/java/org/embulk/spi/util/DynamicPageBuilder.java
+++ b/embulk-core/src/main/java/org/embulk/spi/util/DynamicPageBuilder.java
@@ -14,6 +14,7 @@ import org.embulk.spi.Column;
 import org.embulk.spi.PageBuilder;
 import org.embulk.spi.PageOutput;
 import org.embulk.spi.Schema;
+import org.embulk.spi.time.TimeZoneIds;
 import org.embulk.spi.util.dynamic.SkipColumnSetter;
 
 public class DynamicPageBuilder implements AutoCloseable {
@@ -30,9 +31,13 @@ public class DynamicPageBuilder implements AutoCloseable {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("default_timezone")
-        @ConfigDefault("\"UTC\"")
-        public org.joda.time.DateTimeZone getDefaultTimeZone();
+        public default org.joda.time.DateTimeZone getDefaultTimeZone() {
+            if (getDefaultTimeZoneId() != null) {
+                return TimeZoneIds.parseJodaDateTimeZone(getDefaultTimeZoneId());
+            } else {
+                return null;
+            }
+        }
 
         @Config("column_options")
         @ConfigDefault("{}")
@@ -49,9 +54,9 @@ public class DynamicPageBuilder implements AutoCloseable {
         // org.embulk.spi.time.TimestampFormat is deprecated, but the getter returns TimestampFormat for compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("timestamp_format")
-        @ConfigDefault("\"%Y-%m-%d %H:%M:%S.%N\"")
-        public org.embulk.spi.time.TimestampFormat getTimestampFormat();
+        public default org.embulk.spi.time.TimestampFormat getTimestampFormat() {
+            return new org.embulk.spi.time.TimestampFormat(getTimestampFormatString());
+        }
 
         @Config("timezone")
         @ConfigDefault("null")
@@ -60,9 +65,13 @@ public class DynamicPageBuilder implements AutoCloseable {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("timezone")
-        @ConfigDefault("null")
-        public Optional<org.joda.time.DateTimeZone> getTimeZone();
+        public default Optional<org.joda.time.DateTimeZone> getTimeZone() {
+            if (getTimeZoneId().isPresent()) {
+                return Optional.of(TimeZoneIds.parseJodaDateTimeZone(getTimeZoneId().get()));
+            } else {
+                return Optional.absent();
+            }
+        }
     }
 
     private DynamicPageBuilder(

--- a/embulk-standards/src/main/java/org/embulk/standards/StdoutOutputPlugin.java
+++ b/embulk-standards/src/main/java/org/embulk/standards/StdoutOutputPlugin.java
@@ -14,6 +14,7 @@ import org.embulk.spi.Page;
 import org.embulk.spi.PageReader;
 import org.embulk.spi.Schema;
 import org.embulk.spi.TransactionalPageOutput;
+import org.embulk.spi.time.TimeZoneIds;
 import org.embulk.spi.util.PagePrinter;
 
 public class StdoutOutputPlugin implements OutputPlugin {

--- a/embulk-standards/src/main/java/org/embulk/standards/StdoutOutputPlugin.java
+++ b/embulk-standards/src/main/java/org/embulk/standards/StdoutOutputPlugin.java
@@ -26,9 +26,13 @@ public class StdoutOutputPlugin implements OutputPlugin {
         // Using Joda-Time is deprecated, but the getter returns org.joda.time.DateTimeZone for plugin compatibility.
         // It won't be removed very soon at least until Embulk v0.10.
         @Deprecated
-        @Config("timezone")
-        @ConfigDefault("\"UTC\"")
-        public org.joda.time.DateTimeZone getTimeZone();
+        public default org.joda.time.DateTimeZone getTimeZone() {
+            if (getTimeZoneId() != null) {
+                return TimeZoneIds.parseJodaDateTimeZone(getTimeZoneId());
+            } else {
+                return null;
+            }
+        }
     }
 
     @Override

